debug_utils/sierra-emu/src/vm/circuit.rs

```diff
@@ -12,8 +12,73 @@ use cairo_lang_sierra::{
     program_registry::ProgramRegistry,
 };
 use num_bigint::{BigInt, BigUint, Sign, ToBigInt};
+use num_integer::{ExtendedGcd, Integer};
+use num_traits::{One, ToPrimitive, Zero};
 use smallvec::smallvec;
 
+fn u384_to_struct(num: BigUint) -> Value {
+    let output_big = num.to_bigint().unwrap();
+
+    let mask: BigInt = BigInt::from_bytes_be(Sign::Plus, &[255; 12]);
+
+    let l0: BigInt = &output_big & &mask;
+    let l1: BigInt = (&output_big >> 96) & &mask;
+    let l2: BigInt = (&output_big >> 192) & &mask;
+    let l3: BigInt = (output_big >> 288) & &mask;
+
+    let range = BigInt::ZERO..(BigInt::from(1) << 96);
+    Value::Struct(vec![
+        Value::BoundedInt {
+            range: range.clone(),
+            value: l0,
+        },
+        Value::BoundedInt {
+            range: range.clone(),
+            value: l1,
+        },
+        Value::BoundedInt {
+            range: range.clone(),
+            value: l2,
+        },
+        Value::BoundedInt { range, value: l3 },
+    ])
+}
+
+fn struct_to_u384(struct_members: Vec<Value>) -> BigUint {
+    let [Value::U128(l0), Value::U128(l1), Value::U128(l2), Value::U128(l3)]: [Value; 4] =
+        struct_members.try_into().unwrap()
+    else {
+        panic!()
+    };
+
+    let l0 = l0.to_le_bytes();
+    let l1 = l1.to_le_bytes();
+    let l2 = l2.to_le_bytes();
+    let l3 = l3.to_le_bytes();
+
+    BigUint::from_bytes_le(&[
+        l0[0], l0[1], l0[2], l0[3], l0[4], l0[5], l0[6], l0[7], l0[8], l0[9], l0[10],
+        l0[11], //
+        l1[0], l1[1], l1[2], l1[3], l1[4], l1[5], l1[6], l1[7], l1[8], l1[9], l1[10],
+        l1[11], //
+        l2[0], l2[1], l2[2], l2[3], l2[4], l2[5], l2[6], l2[7], l2[8], l2[9], l2[10],
+        l2[11], //
+        l3[0], l3[1], l3[2], l3[3], l3[4], l3[5], l3[6], l3[7], l3[8], l3[9], l3[10],
+        l3[11], //
+    ])
+}
+
+fn find_nullifier(num: &BigUint, modulus: &BigUint) -> BigUint {
+    let ExtendedGcd { gcd, .. } = num
+        .to_bigint()
+        .unwrap()
+        .extended_gcd(&modulus.to_bigint().unwrap());
+    let gcd = gcd.to_biguint().unwrap();
+
+    // If there's no inverse, find the value which nullifys the operation
```

Comment:
What is a nullifier?

```suggestion
    // If there's no inverse, find the value which nulifies the operation
```

Reply:
If the value is non-invertible a nullifier returned instead. It's a value which satisfies the equation:
`value * nullifier ≡ 0(modulus)`